packages/paste-website/src/pages/blog/2025-03-20-css-variables-ssr-ssg.mdx

```diff
@@ -0,0 +1,189 @@
+export const meta = {
+  title: 'Theme switching with SSR and SSG',
+  slug: '/blog/2025-03-20-css-variables-ssr-ssg/',
+  date: '2025-03-20',
+  author: 'Paste',
+  avatar: 'https://avatars.githubusercontent.com/t/3335688?s=280&v=4',
+  excerpt: 'This blog describes how the Paste website uses CSS variables to switch themes with server-side rendering (SSR) and static site generation (SSG) with our existing ThemeProvider component.',
+};
+
+import Image from 'next/image';
+import {Paragraph} from '@twilio-paste/paragraph';
+import {Box} from '@twilio-paste/box';
+import {InlineCode} from '@twilio-paste/inline-code';
+import {Anchor} from '@twilio-paste/anchor';
+import {CodeBlock, CodeBlockHeader, CodeBlockWrapper} from '@twilio-paste/code-block';
+import {ResponsiveImage} from '../../components/ResponsiveImage';
+
+import DefaultLayout from '../../layouts/DefaultLayout';
+import {getNavigationData} from '../../utils/api';
+
+export default DefaultLayout;
+
+export const getStaticProps = async () => {
+  const navigationData = await getNavigationData();
+  return {
+    props: {
+      navigationData,
+    },
+  };
+};
+
+<ArticleHeader
+  title={meta.title}
+  description={meta.description}
+  date={meta.date}
+  machineDate={meta.date}
+  author={meta.author}
+  avatar={meta.avatar}
+/>
+
+---
+
+<contentwrapper>
+
+<PageAside data={mdxHeadings} hideFeedback />
+
+<ArticleContent>
+
+## What is SSR and SSG?
+
+Server-Side Rendering (SSR) and Static Site Generation (SSG) are two different rendering strategies used in frameworks like Next.js, each with its own advantages and use cases.
+
+SSR generates pages on the server for every request, ensuring that users always receive the latest data. In Next.js, this is achieved using <InlineCode>getServerSideProps</InlineCode>. This approach is useful for dynamic content that changes frequently, such as personalized dashboards or real-time data feeds. However, because the server must generate the page before sending it to the client, it can be slower than other rendering methods.
+
+SSG, on the other hand, generates pages at build time, meaning the content is pre-rendered and stored as static files. In Next.js, this is done using <InlineCode>getStaticProps</InlineCode>. This method is ideal for pages where content doesn’t change often, such as blogs, marketing pages, or documentation. Since the pages are served as static files, they load much faster than SSR-generated pages.
+
+As our docs site data doesn't change between builds we use SSG to generate our pages to utilize the speed benefits of static site generation.
+
+### Issue with theme switching
+
+When using SSG (Static Site Generation) and SSR (Server-Side Rendering) in Next.js, you may sometimes experience a flashbang effect or theme flickering when loading a page. This happens due to differences in how the initial HTML is generated versus how the React client-side hydration process takes over, recognizing the client preferences. Here’s why:
```

Comment:
suggestion: can you include a gif example of the theme flickering?

Reply:
love being able to see the gif! this is a nit but if you want to add it, it could use a tiny caption underneath, like "when the page loads in dark theme, this flicker happens because blah blah blah"